### PR TITLE
Fix the cast issue with return values in dispatcher for FFI upcall

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes fix the cast issue with the values returned from native2InterpJavaUpcallImpl().

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

File: runtime/vm/UpcallVMHelpers.cpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ extern void c_cInterpreter(J9VMThread *currentThread);
 extern bool buildCallInStackFrameHelper(J9VMThread *currentThread, J9VMEntryLocalStorage *newELS, bool returnsObject);
 extern void restoreCallInFrameHelper(J9VMThread *currentThread);
 
-static UDATA JNICALL native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer);
+static U_64 JNICALL native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer);
 static J9VMThread * getCurrentThread(J9UpcallMetaData *data, bool *isCurThrdAllocated);
 static void convertUpcallReturnValue(J9UpcallMetaData *data, U_8 returnType, UDATA *returnStorage);
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:43]: The third parameter should be `U_64 *returnStorage`.

File: runtime/vm/UpcallVMHelpers.cpp [hunk 2]
```diff
@@ -121,7 +121,7 @@ native2InterpJavaUpcallF(J9UpcallMetaData *data, void *argsListPointer)
 double JNICALL
 native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
 {
-	UDATA returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
+	U_64 returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
 	/* The value returned from the upcall method is literally the double precision (64bit) IEEE 754 floating-point
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:125]: Please a the missing hyphen in "64-bit".

File: runtime/vm/UpcallVMHelpers.cpp [hunk 3]
```diff
@@ -140,7 +140,7 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
 U_8 * JNICALL
 native2InterpJavaUpcallStruct(J9UpcallMetaData *data, void *argsListPointer)
 {
-	UDATA returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
+	U_64 returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
 	return (U_8 *)returnValue;
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:144]: This should cast through `UDATA` so we aren't casting to a pointer of a different size:
```
	return (U_8 *)(UDATA)returnValue;
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Other comments/questions:

* in `native2InterpJavaUpcallImpl`
  -  `returnStorage` should be of type `U_64`
  - line 244 can be simplified to `returnsObject = (J9NtcPointer == returnType) || (J9NtcStruct == returnType);`
  - lines 297-299: why is the shift only applied for float?
  - please fix the comment on line 300 which should be `/* !defined(J9VM_ENV_LITTLE_ENDIAN)  */`
  - please fix the indentation on lines 337-339
  
* please add the missing space before `(` on line 426

File: runtime/vm/UpcallVMHelpers.cpp (hunk 1 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: Fixed.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 2 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: Fixed.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 3 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: Fixed.

**Comment by ChengJin01** [comment]
>        returnStorage should be of type U_64

Updated at all places using `returnStorage`.

**Comment by ChengJin01** [comment]
> lines 297-299: why is the shift only applied for float?

We have java code to specially handle `float` by converting to `int` at first and further to `long` in downcall handler at https://github.com/eclipse-openj9/openj9/blob/f3810dfb532be71f8e5a8ced71affe4a6a998181/jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java#L191, in which case we need to address this case in native before passing it on the java stack in upcall on the Big-Endian(BE) platforms.

**Comment by ChengJin01** [comment]
> please fix the comment on line 300 which should be /* !defined(J9VM_ENV_LITTLE_ENDIAN) */

Fixed.

File: runtime/vm/UpcallVMHelpers.cpp [hunk 4]
```diff
@@ -241,7 +241,7 @@ native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
 
 	VM_VMAccess::inlineEnterVMFromJNI(currentThread);
 	returnType = getReturnTypeFromMetaData(data);
-	returnsObject = ((J9NtcPointer == returnType) || (J9NtcStruct == returnType)) ? true : false;
+	returnsObject = ((J9NtcPointer == returnType) || (J9NtcStruct == returnType));
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:244]: The outermost parentheses are redundant.

File: runtime/vm/UpcallVMHelpers.cpp [hunk 5]
```diff
@@ -292,12 +292,14 @@ native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
...
 #if !defined(J9VM_ENV_LITTLE_ENDIAN)
 				/* Right shift the 64-bit float argument by 4 bytes(32 bits) given the actual value
-				 * is placed on the higher 4 bytes on the Big-Endian(BE) platforms.
+				 * is placed on the higher 4 bytes on the Big-Endian(BE) platforms after being converted
+				 * to long via floatToLongArg() in the downcall handler.
+				 * See InternalDowncallHandler.java for details.
 				 */
 				if (J9_FFI_UPCALL_SIG_TYPE_FLOAT == argSigType) {
 					argValue = argValue >> J9_FFI_UPCALL_SIG_TYPE_32_BIT;
 				}
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:301]: I think the shift is required for all 32-bit types (_except_ perhaps float). A 64-bit value was read above on line 292. If only the first 32-bits are wanted, they are the top 32 bits of are what is needed on a big-endian platform unless, in the case of float, the value was first extended to 64 bits.

File: runtime/vm/UpcallVMHelpers.cpp [hunk 6]
```diff
@@ -373,7 +375,7 @@ native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
 	}
 
 	if (!throwOOM) {
-		returnStorage = currentThread->returnValue;
+		returnStorage = (U_64)currentThread->returnValue;
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:378]: This needs to read 64 bits even if `returnValue` is narrower than that. The line should be
```
		/* Read returnStorage from returnValue (and returnValue2 on 32-bit platforms). */
		returnStorage = *(U_64 *)&currentThread->returnValue;
```

File: runtime/vm/UpcallVMHelpers.cpp [hunk 7]
```diff
@@ -452,7 +454,7 @@ getCurrentThread(J9UpcallMetaData *data, bool *isCurThrdAllocated)
  * @param returnStorage a pointer to the return value
  */
 static void
-convertUpcallReturnValue(J9UpcallMetaData *data, U_8 returnType, UDATA *returnStorage)
+convertUpcallReturnValue(J9UpcallMetaData *data, U_8 returnType, U_64 *returnStorage)
```
Comment (keithc-ca) [inline runtime/vm/UpcallVMHelpers.cpp:457]: Please fix the `#endif` comment on line 473 below.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/vm/UpcallVMHelpers.cpp (hunk 4 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: Fixed.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 5 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: I just removed the check on `float` and will need to verify on the supported platforms (specially on AIX and zLinux) to see whether it works good as expected.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 6 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: Updated as suggested above.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 7 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp:455]: Fixed.

**Comment by keithc-ca** [comment]
I think this has reached a good place. Please let me know whether your testing confirms that.

File: runtime/vm/UpcallVMHelpers.cpp (hunk 5 above)
Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp]: The shift doesn't work for other 32bit types on the Big-Endian platforms (e.g. AIX or zLinux) except float in that float is passed in the lower address 4-byte of the slot, while char/short/int are effectively sign-extended to int64 before passing (so they are treated as if they were int64 (confirmed by @zl-wang), which follows what is passed in interpreter.  So we still need to type check for `float` to do shift.

Comment (ChengJin01) [reply runtime/vm/UpcallVMHelpers.cpp:301]: I just restored the type check for `float` in such  case.

**Comment by keithc-ca** [comment]
Can you fix these minor things:
* the missing period on lines 165, 207, 219, 400, 419, 449 & 616
* the indentation of lines 320-322

**Comment by ChengJin01** [comment]
>  the missing period on lines 165, 207, 219, 400, 419, 449 & 616
> the indentation of lines 320-322

Fixed.

**Comment by keithc-ca** [comment]
Thanks for fixing the other minor issues I missed. ;-)

**Comment by keithc-ca** [comment]
jenkins test sanity alinux64 jdk19

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**